agent/agent.c

```diff
@@ -23,39 +23,175 @@
 
 #include <glib-unix.h>
 
+#include <hwangsae/dbus/manager-generated.h>
+#include <hwangsae/dbus/edge-interface-generated.h>
+#include <chamge/chamge.h>
+
+#define DEFAULT_HUB_UID        "abc-987-123"
+#define DEFAULT_BACKEND         CHAMGE_BACKEND_AMQP
+#define DEFAULT_URI            "srt://127.0.0.1:8888"
+
 struct _HwangsaeAgent
 {
   GApplication parent;
 
   HwangsaeRelay *relay;
+  Hwangsae1DBusManager *manager;
+  Hwangsae1DBusEdgeInterface *edge_interface;
+  ChamgeHub *chamge_hub;
 };
 
 /* *INDENT-OFF* */
 G_DEFINE_TYPE (HwangsaeAgent, hwangsae_agent, G_TYPE_APPLICATION)
 /* *INDENT-ON* */
 
+static gboolean
+hwangsae_agent_dbus_register (GApplication * app,
+    GDBusConnection * connection, const gchar * object_path, GError ** error)
+{
+  gboolean ret;
+  HwangsaeAgent *self = HWANGSAE_AGENT (app);
+
+  g_debug ("hwangsae_agent_dbus_register");
+
+  g_application_hold (app);
+
+  /* chain up */
+  ret =
+      G_APPLICATION_CLASS (hwangsae_agent_parent_class)->dbus_register
+      (app, connection, object_path, error);
+
+  if (ret &&
+      !g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON
+          (self->manager), connection,
+          "/org/hwangsaeul/Hwangsae1/Manager", error)) {
+    g_warning ("Failed to export Hwangsae1 D-Bus interface (reason: %s)",
+        (*error)->message);
+  }
+
+  if (ret &&
+      !g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON
+          (self->edge_interface), connection,
+          "/org/hwangsaeul/Hwangsae1/EdgeInterface", error)) {
+    g_warning ("Failed to export Hwangsae1 D-Bus interface (reason: %s)",
+        (*error)->message);
+  }
+
+  return ret;
+}
+
+static void
+hwangsae_agent_dbus_unregister (GApplication * app,
+    GDBusConnection * connection, const gchar * object_path)
+{
+  HwangsaeAgent *self = HWANGSAE_AGENT (app);
+
+  g_debug ("hwangsae_agent_dbus_unregister");
+
+  if (self->manager)
+    g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON
+        (self->manager));
+
+  if (self->edge_interface)
+    g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON
+        (self->edge_interface));
+
+  g_application_release (app);
+
+  /* chain up */
+  G_APPLICATION_CLASS (hwangsae_agent_parent_class)->dbus_unregister (app,
+      connection, object_path);
+}
+
+gboolean
+hwangsae_agent_edge_interface_handle_start (Hwangsae1DBusEdgeInterface * object,
+    GDBusMethodInvocation * invocation, gchar * arg_id, gpointer user_data)
+{
+  ChamgeReturn ret = CHAMGE_RETURN_OK;
+  g_autofree gchar *cmd = NULL;
+  g_autofree gchar *response = NULL;
+  GError *error = NULL;
+  HwangsaeAgent *self = (HwangsaeAgent *) user_data;
+  gchar *uid = NULL;
+
+  cmd =
+      g_strdup_printf ("{\"to\":\"%s\",\"method\":\"streamingStart\", "
```

Comment:
@justinjoy Should we also pass encoding parameters (`width`, `height`, `fps`, `bitrates`)?

I expect the REST API to be just a wrapper around the DBus interfaces and the specification of `POST api/v1.0/srt/start/{edgeId}` has those.